core/src/main/java/io/questdb/griffin/DatabaseSnapshotAgent.java

```diff
@@ -225,10 +222,13 @@ public void completeSnapshot() throws SqlException {
             // Delete snapshot/db directory.
             path.of(configuration.getSnapshotRoot()).concat(configuration.getDbDirectory()).$();
             ff.rmdir(path); // it's fine to ignore errors here
-
             // Release locked readers if any.
             unsafeReleaseReaders();
         } finally {
+            // Resume the WalPurgeJob
+            if (walPurgeJobRunLock != null && walPurgeJobRunLock.isHeldByCurrentThread()) {
```

Comment:
There is no guarantee that `SNAPSHOT PREPARE;` and `SNAPSHOT COMPLETE;` will be called on the same thread, so we may end up in situation when we keep the purge job blocked after `SNAPSHOT COMPLETE;`. To fix that we could use a spin lock which doesn't enforce lock ownership (say, `SimpleSpinLock`, but we need to implement the `tryLock()` method there).

WDYT?

Reply:
Ok, Java's standard lock primitives aren't a good match for a problem.
`ReentrantLock` is owned and not "transferable" between threads ("`snapshot prepare`" and "`snapshot complete`" from different threads). 
`Semaphore` is transferable and not owned, but can have tickets imbalance (`release` method semantic).
Unfortunately, spin-lock doesn't solve the problem as well.
Looks like I have to implement custom assymetric Lock/Handle with something like Peterson algo inside.

Reply:
> Unfortunately, spin-lock doesn't solve the problem as well.

Why do you think so? Considering that we use optimistic locking, a spin lock may be fine. Am I missing anything?